vortex-layout/src/layouts/path.rs

```diff
@@ -0,0 +1,324 @@
+        // the current field
+        let mut new_writers = HashMap::with_capacity(self.leaf_writers.len());
+
+        for (field_path, strategy) in &self.leaf_writers {
+            if field_path.starts_with_field(field)
+                && let Some(subpath) = field_path.clone().step_into()
+            {
+                new_writers.insert(subpath, strategy.clone());
+            }
+        }
+
+        Self {
+            leaf_writers: new_writers,
+            validity: self.validity.clone(),
+            fallback: self.fallback.clone(),
+        }
+    }
+}
+
+/// Specialized strategy for when we exactly know the input schema.
+#[async_trait]
+impl LayoutStrategy for PathStrategy {
+    async fn write_stream(
+        &self,
+        ctx: ArrayContext,
+        segment_sink: SegmentSinkRef,
+        stream: SendableSequentialStream,
+        mut eof: SequencePointer,
+        handle: Handle,
+    ) -> VortexResult<LayoutRef> {
+        let dtype = stream.dtype().clone();
+
+        // Fallback: if the array is not a struct, fallback to writing a single array.
+        if !dtype.is_struct() {
+            return self
+                .fallback
+                .write_stream(ctx, segment_sink, stream, eof, handle)
+                .await;
+        }
+
+        let struct_dtype = dtype.as_struct_fields();
+
+        // Check for unique field names at write time.
+        if HashSet::<_, DefaultHashBuilder>::from_iter(struct_dtype.names().iter()).len()
+            != struct_dtype.names().len()
+        {
+            vortex_bail!("StructLayout must have unique field names");
+        }
+        let is_nullable = dtype.is_nullable();
+
+        // Optimization: when there are no fields, don't spawn any work and just write a trivial
+        // StructLayout.
+        if struct_dtype.nfields() == 0 && !is_nullable {
+            let row_count = stream
+                .try_fold(
+                    0u64,
+                    |acc, (_, arr)| async move { Ok(acc + arr.len() as u64) },
+                )
+                .await?;
+            return Ok(StructLayout::new(row_count, dtype, vec![]).into_layout());
+        }
+
+        // stream<struct_chunk> -> stream<vec<column_chunk>>
+        let columns_vec_stream = stream.map(move |chunk| {
+            let (sequence_id, chunk) = chunk?;
+            let mut sequence_pointer = sequence_id.descend();
+            let struct_chunk = chunk.to_struct();
+            let mut columns: Vec<(SequenceId, ArrayRef)> = Vec::new();
+            if is_nullable {
+                columns.push((
+                    sequence_pointer.advance(),
+                    chunk.validity_mask().into_array(),
+                ));
+            }
+
+            columns.extend(
+                struct_chunk
+                    .fields()
+                    .iter()
+                    .map(|field| (sequence_pointer.advance(), field.to_array())),
+            );
+
+            Ok(columns)
+        });
```

Comment:
Is all this code just the same as the current `StructStrategy`? Is there a reason to maintain both?